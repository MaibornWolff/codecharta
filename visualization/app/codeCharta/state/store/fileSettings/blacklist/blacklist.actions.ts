import { BlacklistItem, CCAction } from "../../../../codeCharta.model"

export enum BlacklistActions {
	ADD_BLACKLIST_ITEM = "ADD_BLACKLIST_ITEM",
	REMOVE_BLACKLIST_ITEM = "REMOVE_BLACKLIST_ITEM",
	SET_BLACKLIST = "SET_BLACKLIST"
}

export interface AddBlacklistAction extends CCAction {
	type: BlacklistActions.ADD_BLACKLIST_ITEM
	payload: BlacklistItem
}

export interface RemoveBlacklistAction extends CCAction {
	type: BlacklistActions.REMOVE_BLACKLIST_ITEM
	payload: BlacklistItem
}

export interface SetBlacklistAction extends CCAction {
	type: BlacklistActions.SET_BLACKLIST
	payload: BlacklistItem[]
}

export type BlacklistAction = AddBlacklistAction | RemoveBlacklistAction | SetBlacklistAction

export function addBlacklistItem(item: BlacklistItem): AddBlacklistAction {
	return {
		type: BlacklistActions.ADD_BLACKLIST_ITEM,
		payload: item
	}
}

export function removeBlacklistItem(item: BlacklistItem): RemoveBlacklistAction {
	return {
		type: BlacklistActions.REMOVE_BLACKLIST_ITEM,
		payload: item
	}
}

export function setBlacklist(blacklist: BlacklistItem[] = defaultBlacklist): SetBlacklistAction {
	return {
		type: BlacklistActions.SET_BLACKLIST,
		payload: blacklist
	}
}

export const defaultBlacklist: BlacklistItem[] = []
