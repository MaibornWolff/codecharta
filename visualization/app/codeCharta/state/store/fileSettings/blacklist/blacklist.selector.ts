import { createSelector } from "@ngrx/store"
import { fileSettingsSelector } from "../fileSettings.selector"

export const blacklistSelector = createSelector(fileSettingsSelector, fileSettings => fileSettings.blacklist)
