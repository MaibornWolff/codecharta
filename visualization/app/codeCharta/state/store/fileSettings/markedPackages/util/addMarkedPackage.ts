import { MarkedPackage } from "../../../../../codeCharta.model"
import { getParent } from "../../../../../util/nodePathHelper"

export const addMarkedPackage = (markedPackagesMap: Map<string, MarkedPackage>, { path, color }: MarkedPackage) => {
	const directMarkedParentPackage = getParent(markedPackagesMap, path)

	if (!directMarkedParentPackage || directMarkedParentPackage.color !== color) {
		markedPackagesMap.set(path, {
			path,
			color
		})
	}

	for (const [key, markedPackage] of markedPackagesMap) {
		if (markedPackage.path === path) {
			if (markedPackage.color !== color) {
				markedPackagesMap.delete(key)
			}
		} else if (markedPackage.path.startsWith(path)) {
			// Remove marked packages with color identical to their parent marked package
			const markedPackageParent = getParent(markedPackagesMap, markedPackage.path)
			if (markedPackageParent && markedPackageParent.color === markedPackage.color) {
				markedPackagesMap.delete(key)
			}
		}
	}
}
