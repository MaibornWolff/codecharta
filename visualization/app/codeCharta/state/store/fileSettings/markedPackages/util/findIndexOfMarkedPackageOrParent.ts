import { MarkedPackage } from "../../../../../codeCharta.model"

/** @return index of exact match if there is an exact match. Otherwise return index of parent or -1  */
export const findIndexOfMarkedPackageOrParent = (markedPackages: MarkedPackage[], nodePath: string) => {
	const indexOfNodePath = markedPackages.findIndex(mp => mp.path === nodePath)
	if (indexOfNodePath !== -1) return indexOfNodePath

	let indexOfNearestParent = -1
	for (let loopIndex = 0; loopIndex < markedPackages.length; loopIndex++) {
		const markedPackage = markedPackages[loopIndex]
		if (
			nodePath.startsWith(markedPackage.path) &&
			(indexOfNearestParent === -1 || markedPackages[indexOfNearestParent].path.length < markedPackage.path.length)
		)
			indexOfNearestParent = loopIndex
	}

	return indexOfNearestParent
}
