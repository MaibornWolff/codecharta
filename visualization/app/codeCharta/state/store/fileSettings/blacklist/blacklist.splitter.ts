import { BlacklistItem } from "../../../../codeCharta.model"
import { BlacklistAction, setBlacklist } from "./blacklist.actions"

export function splitBlacklistAction(payload: BlacklistItem[]): BlacklistAction {
	return setBlacklist(payload)
}
