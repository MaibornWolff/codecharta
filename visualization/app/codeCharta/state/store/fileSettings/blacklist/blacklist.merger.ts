import { BlacklistItem, CCFile } from "../../../../codeCharta.model"
import { getUpdatedBlacklistItemPath } from "../../../../util/nodePathHelper"

export function getMergedBlacklist(inputFiles: CCFile[], withUpdatedPath: boolean): BlacklistItem[] {
	let blacklist: BlacklistItem[] = []

	if (inputFiles.length == 1) {
		return inputFiles[0].settings.fileSettings.blacklist
	}

	for (let inputFile of inputFiles) {
		if (inputFile.settings.fileSettings.blacklist) {
			for (let oldBlacklistItem of inputFile.settings.fileSettings.blacklist) {
				let blacklistItem: BlacklistItem = {
					path: withUpdatedPath
						? getUpdatedBlacklistItemPath(inputFile.fileMeta.fileName, oldBlacklistItem.path)
						: oldBlacklistItem.path,
					type: oldBlacklistItem.type
				}
				const equalBlacklistItems = blacklist.find(b => b.path == blacklistItem.path && b.type == blacklistItem.type)

				if (!equalBlacklistItems) {
					blacklist.push(blacklistItem)
				}
			}
		}
	}
	return blacklist
}
