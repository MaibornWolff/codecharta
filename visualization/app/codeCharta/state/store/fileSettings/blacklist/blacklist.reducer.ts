import { BlacklistItem } from "../../../../codeCharta.model"
import { BlacklistAction, BlacklistActions, setBlacklist } from "./blacklist.actions"
import { addItemToArray, removeItemFromArray } from "../../../../util/reduxHelper"
import { clone } from "../../../../util/clone"

export function blacklist(state: BlacklistItem[] = setBlacklist().payload, action: BlacklistAction): BlacklistItem[] {
	switch (action.type) {
		case BlacklistActions.ADD_BLACKLIST_ITEM:
			return addItemToArray(state, action.payload)
		case BlacklistActions.REMOVE_BLACKLIST_ITEM:
			return removeItemFromArray(state, action.payload)
		case BlacklistActions.SET_BLACKLIST:
			return clone(action.payload)
		default:
			return state
	}
}
