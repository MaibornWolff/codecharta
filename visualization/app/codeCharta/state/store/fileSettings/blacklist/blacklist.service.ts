import { StoreService, StoreSubscriber } from "../../../store.service"
import { IRootScopeService } from "angular"
import { BlacklistItem, FileState } from "../../../../codeCharta.model"
import { BlacklistActions, setBlacklist } from "./blacklist.actions"
import _ from "lodash"
import { getMergedBlacklist } from "./blacklist.merger"
import { FileStateService, FileStateSubscriber } from "../../../fileState.service"
import { FileStateHelper } from "../../../../util/fileStateHelper"

export interface BlacklistSubscriber {
	onBlacklistChanged(blacklist: BlacklistItem[])
}

export class BlacklistService implements StoreSubscriber, FileStateSubscriber {
	private static BLACKLIST_CHANGED_EVENT = "blacklist-changed"

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		StoreService.subscribe(this.$rootScope, this)
		FileStateService.subscribe(this.$rootScope, this)
	}

	public onStoreChanged(actionType) {
		if (_.values(BlacklistActions).includes(actionType)) {
			this.notify(this.select())
		}
	}

	public onFileStatesChanged(fileStates: FileState[]) {
		this.merge(fileStates)
	}

	private merge(fileStates: FileState[]) {
		const visibleFiles = FileStateHelper.getVisibleFileStates(fileStates).map(x => x.file)
		const withUpdatedPath = FileStateHelper.isPartialState(fileStates)
		const newBlacklist = getMergedBlacklist(visibleFiles, withUpdatedPath)
		this.storeService.dispatch(setBlacklist(newBlacklist))
	}

	private select() {
		return this.storeService.getState().fileSettings.blacklist
	}

	private notify(newState: BlacklistItem[]) {
		this.$rootScope.$broadcast(BlacklistService.BLACKLIST_CHANGED_EVENT, { blacklist: newState })
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: BlacklistSubscriber) {
		$rootScope.$on(BlacklistService.BLACKLIST_CHANGED_EVENT, (event, data) => {
			subscriber.onBlacklistChanged(data.blacklist)
		})
	}
}
