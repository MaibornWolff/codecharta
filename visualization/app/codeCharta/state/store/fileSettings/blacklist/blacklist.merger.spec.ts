import { BlacklistItem, BlacklistType, CCFile } from "../../../../codeCharta.model"
import { getMergedBlacklist } from "./blacklist.merger"
import { TEST_FILE_DATA } from "../../../../util/dataMocks"
import _ from "lodash"

describe("BlacklistMerger", () => {
	let file1: CCFile
	let file2: CCFile

	beforeEach(() => {
		file1 = _.cloneDeep(TEST_FILE_DATA)
		file1.fileMeta.fileName = "file1"

		file2 = _.cloneDeep(TEST_FILE_DATA)
		file2.fileMeta.fileName = "file2"
	})

	describe("getMergedBlacklist", () => {
		const blacklistItem1: BlacklistItem = { path: "/root/nodeA", type: BlacklistType.exclude }
		const blacklistItem2: BlacklistItem = { path: "/another/nodeB", type: BlacklistType.flatten }
		const blacklistItem3: BlacklistItem = { path: "/another/nodeC", type: BlacklistType.exclude }
		const blacklistItem4: BlacklistItem = { path: "*prefix/nodeD", type: BlacklistType.flatten }
		const blacklistItem1Duplicate: BlacklistItem = { path: "/root/nodeA", type: BlacklistType.exclude }

		it("should merge blacklist for different paths", () => {
			file1.settings.fileSettings.blacklist = [blacklistItem1, blacklistItem2]
			file2.settings.fileSettings.blacklist = [blacklistItem3, blacklistItem4]
			expect(getMergedBlacklist([file1, file2], false)).toMatchSnapshot()
		})

		it("should only contain unique paths+type", () => {
			file1.settings.fileSettings.blacklist = [blacklistItem1, blacklistItem2]
			file2.settings.fileSettings.blacklist = [blacklistItem1Duplicate, blacklistItem4]
			expect(getMergedBlacklist([file1, file2], false)).toMatchSnapshot()
		})
	})
})
