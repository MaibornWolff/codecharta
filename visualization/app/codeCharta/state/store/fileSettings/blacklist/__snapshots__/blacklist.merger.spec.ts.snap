// Jest Snapshot v1, https://goo.gl/fbAQLP


exports[`BlacklistMerger getMergedBlacklist should merge blacklist for different paths 1`] = `
Array [
  Object {
    "path": "/root/nodeA",
    "type": "exclude",
  },
  Object {
    "path": "/another/nodeB",
    "type": "flatten",
  },
  Object {
    "path": "/another/nodeC",
    "type": "exclude",
  },
  Object {
    "path": "*prefix/nodeD",
    "type": "flatten",
  },
]
`;

exports[`BlacklistMerger getMergedBlacklist should only contain unique paths+type 1`] = `
Array [
  Object {
    "path": "/root/nodeA",
    "type": "exclude",
  },
  Object {
    "path": "/another/nodeB",
    "type": "flatten",
  },
  Object {
    "path": "*prefix/nodeD",
    "type": "flatten",
  },
]
`;